doc/source/virtual-cluster/design-overview.rst

```diff
@@ -0,0 +1,168 @@
+   - The expected replica sets specified in the request have to be valid.
+
+2. Select enough unassigned nodes from the primary cluster.
+3. Remove the selected nodes from the primary cluster, and add them to the new virtual cluster.
+4. Flush the metadata of the newly created virtual cluster to Redis and publish it to all raylets.
+
+**Update:**
+
+When updating an existing virtual cluster, GCS approximately goes through the following steps:
+
+1. Pre-check:
+
+   - The request has to specify a valid and existing virtual cluster ID.
+   - The expected replica sets specified in the request have to be valid.
+   - The revision specified in the request has to be the latest one.
+   - The request can not change the virtual cluster from divisible to indivisible (or vice-versa) on the fly.
+
+2. Select enough unassigned nodes from the primary cluster (if the virtual cluster expects to scale up), and select enough idle nodes to be removed from the virtual cluster (if the virtual cluster expects to scale down).
+3. Update the node type and count of the primary cluster and the virtual cluster.
+4. Flush the metadata of the updated virtual cluster to Redis and publish it to all raylets.
+
+Note, currently we do not support updating a job cluster on the fly. It is our plan to make it available in the next release.
+
+**Remove:**
+
+When removing a virtual cluster, GCS approximately goes through the following steps:
+
+1. Pre-check: The request has to specify a valid and existing virtual cluster ID.
+2. If there is any jobs/actors/tasks/placement groups still running inside the virtual cluster, then this request should fail.
+3. Return the virtual cluster's nodes back to its father cluster.
+4. Flush the metadata of the newly removed virtual cluster to Redis and publish it to all raylets.
+
+.. _virtual-cluster-fault-tolerance:
+
+Virtual Cluster Fault Tolerance
+-------------------------------
+
+Let nodes inside logical indivisible cluster be :math:`V_{indivisble}` , nodes inside divisible cluster be :math:`V_{divisible}`, nodes inside job cluster be :math:`V_{job}`, undivided nodes inside divisible cluster be :math:`V_{undivided}`, then:
+- For divisible cluster: :math:`V_{divisible} = V_{undivided} + \sum{V_{job}}`, :math:`V_{job}` is divided from :math:`V_{divisible}`
+- For indivisible cluster: :math:`V_{indivisible} = V_{undivided}`, it means that the indivisble cluster can not be diveded any more and all nodes inside indivisible cluster are undivieded.
+
+**Raylet Failover**
+When Rayle dies or restarts, the GCS Server will detect that the node is dead and then update the status of the corresponding node in the virtual cluster, and the virtual cluster manager will periodically check for dead nodes and replace them with new ones.
+
+There are 4 scenario that the virtual cluster manager will replenish new nodes for the virtual cluster, and the state of the virtual cluster need flush or publish:
```

Comment:
There are four scenarios in which the virtual cluster manager will replenish new nodes for the virtual cluster, and the state of the virtual cluster needs to be flushed or published: